### PR TITLE
MOSFiT 2.0

### PR DESCRIPTION
**Pull request by guillochon**
## Summary

This is **MOSFiT 2.0**: a packaging-and-runtime overhaul on top of current `master`. Development installs move to **uv** and `pyproject.toml`; the default sampler is **dynesty 3.1**; torch is optional; likelihoods can use a local process pool; several hot-path modules are vectorized without changing the math. The TDE viscous delay is now a jitted closed-form recurrence instead of a 1000-point trapezoid (**~500×** on that kernel, **~3.3×** on a full PS1-10jh `ln_likelihood`).

### Packaging and Python

- Replace `setup.py` / `setup.cfg` / `MANIFEST.in` as the source of truth with **hatchling** + `pyproject.toml` + `uv.lock`. Version is **2.0.0**.
- **`requires-python = ">=3.11"`** (classifiers 3.11–3.14). Local `.python-version` pins **3.14.7** for uv; any 3.11+ interpreter is allowed.
- Runtime floor is driven by **astrocats 0.5.0** (NumPy 2.3 / Astropy 7.1). Drop the vendored astrocats 0.3.37 wheel. `astrocats>=0.5.0` resolves from **PyPI**.
- Optional extras: `mpi` (mpi4py), `sedona` (PyTorch, SESN SEDONA emulator only), `docs` (Sphinx).
- `import mosfit` does **not** import torch. SED modules lazy-load; `sesn_sedona` tells the user to `uv sync --extra sedona`.
- NumPy 2: generative dummy times (`max_time` default) no longer pass an empty string into `linspace`.
- Docker: `python:3.14.7-slim`, `uv sync --frozen --extra mpi`, no torch. Smoke-check that `import mosfit` does not pull in torch. `.dockerignore` added.
- conda-forge: hatchling sdist (no `setup.py`, no required PyTorch). Template in `recipe/`. **astrocats >= 0.5.0** must land on conda-forge before the mosfit feedstock can solve (channel still has 0.3.37).

### Sampling

- Default `-D` / Python `method` is **`dynesty`** (was ensembler). `ensembler` and `ultranest` remain selectable.
- Require **dynesty >= 3.1** (`evid_thresh` / iterator attributes). Nested live points are `nlive = 20 * ndim`; `--num-walkers` does **not** set nlive.
- `-R` is remaining evidence (`dlogz`, default `0.02`) for dynesty; PSRF still applies to ensembler.
- Nested sampling: do not rebuild `sampler.results` every step; disable unused bound history (`save_bounds=False`); snapshot Results once after baseline and after each `combine_runs`. CPU lulls between batches (ellipsoid rebuilds on the parent) are expected.

### Parallelism and likelihood speed

- **`--max-cores N`**: local `schwimmbad` process pool (Windows spawn). Default 1. `mpirun` still takes precedence.
- Vectorize SED/photometry by **unique band** (blackbody, losextinction, photometry). NumPy residuals in diagonal/kernel.
- Photometry: interpolate filter curves onto `sample_wavelengths` **once** per sample grid (same integrals).
- Keep SEDs as rectangular `(n_obs, n_wav)` float64 arrays instead of object-dtype row lists (no `np.stack` round trip through photometry). Isolated blackbody + photometry-like path is **1.85×**.
- TDE `Fallback`: cache Tout coefficients, `np.interp` instead of `interp1d`, NumPy NaN/array assembly. Golden draws match the previous engine.
- **Viscous delay:** replace the 1000-point log-spaced trapezoid with a Numba piecewise-linear exponential recurrence (exact integral of the same interpolant). Isolated kernel **~526×**; PS1-10jh `ln_likelihood` **~3.3×**.

#### Viscous exponential recurrence

The old viscous transform spent most of a TDE likelihood on a `(n_times × ~1000)` interp + `exp` + `trapezoid`. `L(t)` was already piecewise linear on the dense grid; the convolution `(1/τ) ∫ L(s) exp((s−t)/τ) ds` has a closed form on each segment. That walk is JIT-compiled.

PS1-10jh, `x = 0.5`, same machine. Kernel times are 40 repeats after warmup; TDE times are 20 repeats after 3 warmups (post-rectangular-SED baseline **10.035 ms**).

```mermaid
---
config:
  xyChart:
    width: 700
    height: 320
    showLegend: true
  themeVariables:
    xyChart:
      plotColorPalette: "#8b949e, #3fb950"
---
xychart-beta
    title "PS1-10jh TDE ln_likelihood (ms)"
    x-axis ["1000-pt trapz", "Jitted recurrence"]
    y-axis "Mean wall time (ms)" 0 --> 12
    bar "Wall time" [10.04, 3.02]
```

```mermaid
---
config:
  xyChart:
    width: 700
    height: 260
    showLegend: false
  themeVariables:
    xyChart:
      plotColorPalette: "#4493f8"
---
xychart-beta horizontal
    title "Speedup vs 1000-point viscous quadrature"
    x-axis ["Viscous kernel", "TDE ln_likelihood"]
    y-axis "Speedup" 0 --> 550
    bar [526, 3.32]
```

| Path | Before | After | Speedup |
| --- | ---: | ---: | ---: |
| Viscous kernel (PS1-10jh arrays) | 7.22 ms | 0.014 ms | **526×** |
| PS1-10jh TDE `ln_likelihood` | 10.04 ms | 3.02 ms | **3.32×** |

Physics vs the old trapezoid (same piecewise-linear `L`):

- Isolated recurrence vs 1000-pt trapz: max relative **6.5×10⁻⁵** (`τ = 3.5`); **≤ 3.7×10⁻⁴** for `τ` from `10⁻³` to `100`.
- Constant `L` matches `L(1 − e^{−t/τ})` to `1e-12`.
- Restoring the trapz on PS1-10jh recovers lnL **−1032.9597765404137** exactly. The recurrence is **−1032.7869483449351** (Δ **+0.173**). Model magnitudes: max **0.83 mmag**, median **0.28 mmag**.

The millimag shift is the old rule not being fully converged, not a different delay kernel.

#### Rectangular SED arrays

Wall-clock speedup from keeping SEDs as float64 `(n_obs, n_wav)` arrays. Same machine, consecutive runs with the patch stashed then restored. Synthetic path: 400 observations × 25 wavelengths, 6 bands; 5 warmup + 40 timed calls. TDE `ln_likelihood` in this table is **before** the viscous recurrence (viscous still dominated).

```mermaid
---
config:
  xyChart:
    width: 700
    height: 360
    showLegend: true
  themeVariables:
    xyChart:
      plotColorPalette: "#8b949e, #3fb950"
---
xychart-beta
    title "Mean wall time, synthetic SED path (us)"
    x-axis ["Blackbody process", "BB + photometry-like", "Photometry consume"]
    y-axis "Mean wall time (us)" 0 --> 700
    bar "Object-dtype lists (before)" [329, 599, 270]
    bar "Rectangular float64 (after)" [237, 323, 86]
```

```mermaid
---
config:
  xyChart:
    width: 700
    height: 280
    showLegend: false
  themeVariables:
    xyChart:
      plotColorPalette: "#4493f8"
---
xychart-beta horizontal
    title "SED-path speedup vs object-dtype lists (before / after)"
    x-axis ["Blackbody process", "BB + photometry-like", "Photometry consume", "TDE lnL (pre-viscous)"]
    y-axis "Speedup" 0 --> 3.5
    bar [1.39, 1.85, 3.12, 0.99]
```

| Path | Before | After | Speedup |
| --- | ---: | ---: | ---: |
| Blackbody process | 329.3 µs | 236.9 µs | **1.39×** |
| Blackbody + photometry-like | 599.5 µs | 323.4 µs | **1.85×** |
| Photometry-like consume only (`np.stack` vs 2-D slice) | 270.2 µs | 86.5 µs | **3.12×** |
| PS1-10jh TDE `ln_likelihood` (before viscous recurrence) | 9.958 ms | 10.035 ms | **~1.00×** |

The 3.12× gain is the restack this change removes. That was a few hundred microseconds inside a ~10 ms TDE likelihood, so it did not move the full-model number until viscous was replaced.

### Tests, CI, and coverage

GitHub Actions (`.github/workflows/tests.yml`) runs `uv sync --frozen` then `uv run python mosfit/tests/run_all.py` on Ubuntu 3.11 / 3.14 and Windows 3.14, plus a job that **pip-installs the hatchling sdist** (conda-forge shape, no `uv.lock`). That runner discovers every `mosfit/tests/_test_*.py` plus the repo-root generative smoke `test.py`, collects **line coverage**, and fails if coverage drops below **70%**.

Local line coverage of the package (omitting tests and the optional torch SEDONA extra):

- Previous three scripts only (`_test_speedup_numerics.py`, `_test_pool_likelihood.py`, `test.py`): **47%**
- Current suite: **76%**

The obsolete CLI script `test.sh` (catalog-by-name fetch, `--offline`, `-F covariance`, upload) is removed.

#### Coverage for every built-in transient type

Real events covered two types: an SLSN (`LSQ12dlf.json`) and a TDE (`PS1-10jh.json`). Every other built-in model was only ever exercised with two-point dummy photometry, which never touches band resolution, upper limits, host extinction, or k-corrections. The Open Astronomy Catalogs are no longer fetched by name, so real data for the other 19 types was not an option.

`mosfit/tests/events/` now holds one **simulated** event per built-in model (~360 KB for all 20, against 5.3 MB for the single real `LSQ12dlf.json`). Each is a small catalog-format light curve the matching model generated itself: a realistic cadence in **already-cached** PS1 / Swift-UVOT / 2MASS bands, noisy magnitudes, non-detections past a per-instrument limit, and a redshift and E(B-V) chosen for that kind of transient. Nothing is downloaded, so they run wherever the tracked SVO filter cache is present.

`mosfit/tests/make_event_fixtures.py` regenerates them. It searches each model's prior for the draw that most looks like a transient (peak near 19th magnitude, early in the campaign, mostly detected) with a per-model seed, so a full regeneration reproduces every file byte for byte.

One thing this shook out, not fixed here: `mosfit/models/csmni/csmni.json` replaces the base model's `luminositydistance` parameter with a plain fixed one at `1e-05` Mpc. `csmni` therefore reads the event redshift but pins the distance at 10 pc, returning absolute magnitudes where every other model returns apparent ones (`csm` on the same event gives 277.5 Mpc; `csmni` gives `1e-05`). It is the only model that overrides that key. Treating it as a model change rather than a test change, its fixture carries an explicit `lumdist` to work around it, with a comment saying why.

Scripts in the suite:

- `mosfit/tests/_test_speedup_numerics.py`: blackbody, photometry, diagonal, mm83, fallback goldens, viscous recurrence vs old quadrature; `import mosfit` must not import torch
- `mosfit/tests/_test_pool_likelihood.py`: serial vs 10-core TDE `ln_likelihood` on a fixed vector
- `mosfit/tests/_test_lsq12dlf.py`: load `mosfit/tests/LSQ12dlf.json` and evaluate one SLSN likelihood
- `mosfit/tests/_test_models_likelihood.py`: load every packaged model except `sesn_sedona` and evaluate one likelihood
- `mosfit/tests/_test_transient_types.py`: for **every** built-in transient type, resolve its simulated event off disk through the public fetcher, check all of its photometry was ingested, evaluate a likelihood, and run a short ensemble fit. Models are discovered from `mosfit/models`, so a new model without a fixture fails here instead of going untested
- `mosfit/tests/_test_csv_input.py`: the non-JSON input path. Converts `mosfit/tests/events/sim_default.csv` in a temporary directory with `exit_on_prompt` set (any column the converter cannot identify on its own fails the test rather than waiting for a human), checks the conversion reproduces the JSON fixture's photometry, redshift and E(B-V) row for row, and fits the result
- `mosfit/tests/_test_ensembler.py`: short `-D ensembler` smoke on dummy `exppow` data
- `mosfit/tests/_test_modules.py`: direct `process()` coverage for engines, parameters, SEDs, transforms, outputs
- `mosfit/tests/_test_unit.py`: utils, plotting helpers, printer, fetcher (local JSON only), ASCII converter, CLI/`get_parser`
- `mosfit/tests/_test_ptsampler.py`: vendored PTSampler / MOSSampler
- `test.py`: generative `fit_events` for `magni` and `slsn` (`iterations=1`)

Fixtures: `mosfit/tests/PS1-10jh.json` (Open TDE Catalog), `mosfit/tests/LSQ12dlf.json`, and the 20 simulated events plus one CSV mirror under `mosfit/tests/events/`.

Full local suite runtime goes from **2m09s** to **4m42s**.

### Documentation

README and Sphinx updated for Python 3.11+, uv, extras, dynesty default, `--max-cores`, `-R`/`dlogz`, SEDONA extra, Docker, the Python `Fitter` API, and MOSFiT 2.0. Autodoc skips torch-backed SEDONA classes so RTD can build without the `sedona` extra. `docs/models.rst` documents the simulated per-model fixtures and how to regenerate them. Release notes: `CHANGELOG.md` (stage under **Unreleased**, move under a version heading when tagging). conda-forge notes: `recipe/`.

## Breaking / behavioral notes

- **This is MOSFiT 2.0.0.** Python **< 3.11 is unsupported.**
- **Default sampler is dynesty**, not ensemble MCMC. Old scripts that assumed emcee defaults should pass `-D ensembler`.
- **Torch is not a core dependency.** `sesn_sedona` needs `uv sync --extra sedona` or `pip install 'mosfit[sedona]'`.
- Event data remains **local files only** (no catalog download by name); that predates most of this branch but docs now match it.
- Published conda-forge / PyPI wheels may lag this branch until a release. conda-forge also needs **astrocats 0.5.0** on that channel first.

## Test plan

- [x] `uv run python mosfit/tests/run_all.py` (full suite + coverage; local result **76%**, fail-under **70%**)
- [x] `uv run python mosfit/tests/_test_transient_types.py` (all 20 built-in transient types loaded and fit)
- [x] `uv run python mosfit/tests/_test_csv_input.py` (CSV converted, compared against the JSON fixture, and fit)
- [x] `uv run python mosfit/tests/make_event_fixtures.py` reproduces every committed fixture byte for byte
- [ ] GitHub Actions Tests workflow green on Ubuntu 3.11 / 3.14 and Windows 3.14
- [ ] `uv sync` on Python 3.11 and 3.14; `import mosfit` must succeed and must not import torch
- [ ] Short dynesty smoke: `uv run mosfit -e mosfit/tests/PS1-10jh.json -m tde --max-cores 4 --no-fracking -i 1`
- [ ] `uv sync --extra sedona` then load/import the SESN SEDONA emulator
- [x] `uv sync --extra docs` and `sphinx-build docs docs/_build/html`
- [ ] Docker build from repo root; container `import mosfit` without torch
- [ ] Windows: `--max-cores 2` on a small local JSON fit (spawn)

### CODE REVIEW

**Comment by guillochon** [comment]
This should knock out https://github.com/guillochon/MOSFiT/issues/246 and supercede #247 .

File: mosfit/__init__.py [hunk 1]
```diff
@@ -22,27 +22,25 @@
...
-            if int(vparts[0]) < int(vneed[0]):
-                right_astrocats = False
-            elif int(vparts[1]) < int(vneed[1]):
-                right_astrocats = False
-            elif int(vparts[2]) < int(vneed[2]):
-                right_astrocats = False
+vneed = [str(part) for part in _ASTROCATS_MIN_VERSION]
+if int(vparts[0]) < _ASTROCATS_MIN_VERSION[0]:
+    right_astrocats = False
+elif int(vparts[1]) < _ASTROCATS_MIN_VERSION[1]:
```
[the remark this answers is not in the document]
Comment (guillochon) [reply mosfit/__init__.py]: Fixed in 22795c6: `_version_triple` takes the leading X.Y.Z integers from PEP 440 strings (`0.5.0rc1`, `0.5.0.post1`), so `import mosfit` no longer raises `ValueError`.

File: mosfit/modules/observables/photometry.py [hunk 2]
```diff
@@ -92,21 +111,24 @@ def __init__(self, **kwargs):
         bands = listify(bands)
 
         self._dir_path = os.path.dirname(os.path.realpath(__file__))
-        self._filter_run_path = os.path.join('modules','observables')
-        if not os.path.exists(self._filter_run_path):
-            os.makedirs(self._filter_run_path)
+        self._filter_run_path = os.path.join('modules', 'observables')
+        os.makedirs(
+            os.path.join(self._filter_run_path, 'filters'), exist_ok=True)
+        os.makedirs(os.path.join(self._dir_path, 'filters'), exist_ok=True)
```
[the remark this answers is not in the document]
Comment (guillochon) [reply mosfit/modules/observables/photometry.py]: Fixed in 22795c6: Photometry only creates the CWD SVO cache (`modules/observables/filters`). It no longer mkdirs under the installed package path.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+{% set version = "2.0.0" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://pypi.org/packages/source/{{ name[0] }}/{{ name }}/mosfit-{{ version }}.tar.gz
+  sha256: # replace with the PyPI sdist hash after the GitHub/PyPI release
+
```
[the remark this answers is not in the document]
Comment (guillochon) [reply recipe/meta.yaml:11]: Fixed in 22795c6: `sha256` is now the explicit placeholder `replace-with-pypi-sdist-sha256` instead of YAML null.

**Comment by guillochon** [comment]
Addressed the suppressed Copilot notes from the latest review in 68c8b1c:

- fit_events max_cores: default is now None, so Fitter(max_cores=10).fit_events(...) keeps the constructor pool size unless the caller passes an explicit override.
- CWD module overrides: custom leaves load as modules.<kind>.<leaf> with the process CWD on sys.path (and the same path is inserted before spawning workers). Packaged modules stay mosfit.modules.<kind>.<leaf> so Numba cache and Windows spawn still import them.
